nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/pubsub/PublishGCPubSub.java

```diff
@@ -213,61 +289,139 @@ public List<ConfigVerificationResult> verify(final ProcessContext context, final
+            messageTracker.add(new FlowFileResult(flowFile, futures));
+            getLogger().trace("Parsing of FlowFile (ID:{}) records complete, now {} messages", flowFile.getId(), messageTracker.size());
+        }
+        finishBatch(session, flowFileBatch, stopWatch, messageTracker);
+    }
+
+    private ApiFuture<String> publishOneRecord(
+            final ProcessContext context,
+            final FlowFile flowFile,
+            final ByteArrayOutputStream baos,
+            final RecordSetWriter writer,
+            final Record record) throws IOException {
+        baos.reset();
+        writer.write(record);
+        writer.flush();
+        return publishOneMessage(context, flowFile, baos.toByteArray());
+    }
+
+    private ApiFuture<String> publishOneMessage(final ProcessContext context,
+                                                final FlowFile flowFile,
+                                                final byte[] content) {
+        final PubsubMessage message = PubsubMessage.newBuilder()
+                .setData(ByteString.copyFrom(content))
+                .setPublishTime(Timestamp.newBuilder().build())
+                .putAllAttributes(getDynamicAttributesMap(context, flowFile))
+                .build();
+        return publisher.publish(message);
+    }
+
+    private void finishBatch(final ProcessSession session,
+                             final List<FlowFile> flowFileBatch,
+                             final StopWatch stopWatch,
+                             final MessageTracker messageTracker) {
+        try {
+            getLogger().trace("Submit of batch complete, size {}", messageTracker.size());
+            final List<String> messageIdsSuccess = ApiFutures.successfulAsList(messageTracker.getFutures()).get();
+            getLogger().trace("Send of batch complete, success size {}", messageIdsSuccess.size());
+            messageTracker.reconcile(messageIdsSuccess);
+            final String topicName = publisher.getTopicNameString();
+            for (final FlowFileResult flowFileResult : messageTracker.getFlowFileResults()) {
+                final Map<String, String> attributes = new HashMap<>();
+                //attributes.put(MESSAGE_ID_ATTRIBUTE, messageIdFuture.get());  // what to do if using record strategy?
+                attributes.put(TOPIC_NAME_ATTRIBUTE, topicName);
+                final FlowFile flowFile = session.putAllAttributes(flowFileResult.getFlowFile(), attributes);
+                final String transitUri = String.format(TRANSIT_URI_FORMAT_STRING, topicName);
+                session.getProvenanceReporter().send(flowFile, transitUri, stopWatch.getElapsed(TimeUnit.MILLISECONDS));
+                session.transfer(flowFile, flowFileResult.getRelationship());
+                if (flowFileResult.getException() != null) {
+                    getLogger().error("FlowFile send failure", flowFileResult.getException());

```

Comment:
The FlowFile should be included.
```suggestion
                    getLogger().error("Send Failed for {}", flowFile, flowFileResult.getException());

```